File: mobile/www/js/services/kodik.js
// Источник видео Kodik для серии: нормализованный VideoSource, кэш и отмена.
//
// Kodik отдаёт видео только через свой встраиваемый плеер (публичный iframe-API). Прямые ссылки на поток
// спрятаны внутри плеера и шифруются — разбирать их ненадёжно и против правил Kodik, поэтому iframe остаётся,
// а надёжность — за счёт порядка загрузки (сессии), start_from и понятных ошибок.
import { request } from "../core/api/http.js";
import { NoSourceError } from "../core/errors.js";

const ANIMELIB = "https://api.cdnlibs.org/api";
const ANIMELIB_H = { "Site-Id": "5" };
const PLAYERS_TTL = 600;   // список плееров серии: одна и та же серия при повторном открытии — без запроса

/** Нормализованная ссылка Kodik: https, без своего выбора озвучки и серий (они — в нашем интерфейсе), с позицией. */
export function kodikUrl(src, { startFrom = 0 } = {}) {
  const url = new URL(src.startsWith("//") ? "https:" + src : src);
  url.searchParams.set("translations", "false");
  if (url.pathname.startsWith("/season/") || url.pathname.startsWith("/serial/")) url.searchParams.set("only_episode", "true");
  if (startFrom > 5) url.searchParams.set("start_from", String(Math.floor(startFrom)));
  else url.searchParams.delete("start_from");
  return url.href;
}

/** Плеер Kodik нужной команды из списка плееров серии AnimeLib; нет её — первый доступный (fallback). */
export function pickKodikPlayer(players, teamId) {
  const kodik = (players || []).filter((p) => p.player === "Kodik" && p.src);
  const exact = kodik.filter((p) => p.team?.id === +teamId);
  const chosen = (exact.length ? exact : kodik)[0];
  return chosen ? { src: chosen.src, team: chosen.team?.name || null, fallback: !exact.length } : null;
}

/**
 * @param {import("../domain/models.js").Episode} ep
 * @param {string|number} teamId
 * @param {{signal?: AbortSignal, startFrom?: number, fallbackName?: string}} o
 * @returns {Promise<import("../domain/models.js").VideoSource>}
 */
export async function resolveKodikSource(ep, teamId, { signal, startFrom = 0, fallbackName = null } = {}) {
  if (ep.kodik) return { kind: "iframe", url: kodikUrl(ep.kodik, { startFrom }), team: fallbackName, fallback: false };
  if (!ep.animelib) throw new NoSourceError("no episode id");
  const d = await request(`${ANIMELIB}/episodes/${ep.animelib}`, { headers: ANIMELIB_H, ttl: PLAYERS_TTL, signal });
  const chosen = pickKodikPlayer(d?.data?.players, teamId);
  if (!chosen) throw new NoSourceError("no kodik player");
  return { kind: "iframe", url: kodikUrl(chosen.src, { startFrom }), team: chosen.team, fallback: chosen.fallback };
}

File: anime_app/domain/kodik.py
"""Плеер Kodik: ссылка на серию и выбор плеера команды озвучки. Те же правила — в мобильной версии (services/kodik.js).

Kodik отдаёт видео только через свой встраиваемый плеер (публичный iframe-API), поэтому источник — ссылка на плеер.
"""
from __future__ import annotations

import urllib.parse

MIN_START = 5   # меньше — начинаем сначала (как и прогресс: первые 5 с не записываются)


def kodik_url(src: str, start_from: float = 0) -> str:
    """https, без своего выбора озвучки и серий (они — в нашем интерфейсе), с позицией продолжения."""
    parts = urllib.parse.urlsplit(("https:" + src) if src.startswith("//") else src)
    query = [(k, v) for k, v in urllib.parse.parse_qsl(parts.query, keep_blank_values=True)
             if k not in ("translations", "only_episode", "start_from")]
    query.append(("translations", "false"))
    if parts.path.startswith(("/season/", "/serial/")):
        query.append(("only_episode", "true"))
    if start_from and start_from > MIN_START:
        query.append(("start_from", str(int(start_from))))
    return urllib.parse.urlunsplit((parts.scheme, parts.netloc, parts.path, urllib.parse.urlencode(query), ""))


def pick_kodik_player(players: list[dict] | None, team_id) -> dict | None:
    """Плеер нужной команды из списка плееров серии AnimeLib; нет её — первый доступный (fallback)."""
    kodik = [p for p in players or [] if p.get("player") == "Kodik" and p.get("src")]
    exact = [p for p in kodik if (p.get("team") or {}).get("id") == int(team_id)]
    chosen = (exact or kodik or [None])[0]
    if not chosen:
        return None
    return {"src": chosen["src"], "team": (chosen.get("team") or {}).get("name"), "fallback": not exact}
